### PR TITLE
sasrec tests

### PR DESCRIPTION
**Pull request by spirinamayya**
added sasrec test

### CODE REVIEW

File: rectools/models/sasrec.py [hunk 1]
```diff
@@ -22,47 +22,55 @@
...
-        """TODO"""
+        """Forward pass."""
         raise NotImplementedError()
 
 
 class IdEmbeddingsItemNet(ItemNetBase):
     """
-    Base class for item embeddings. To use more complicated logic then just id embeddings inherit
-    from this class and pass your custom ItemNet to your model params
+    Network for item embeddings.
```
Comment (blondered) [inline rectools/models/sasrec.py:64]:
```suggestion
    Network for item embeddings based only on item ids.
```

File: rectools/models/sasrec.py [hunk 2]
```diff
@@ -111,14 +153,39 @@ def __init__(self, n_factors: int, n_factors_ff: int, dropout_rate: float) -> No
...
+        """
         output = self.ff_relu(self.ff_dropout1(self.ff_linear1(seqs)))
         fin = self.ff_dropout2(self.ff_linear2(output))
         return fin
 
 
 class SASRecTransformerLayers(TransformerLayersBase):
-    """Exactly SASRec authors architecture but with torch MHA realisation"""
+    """
+    Exactly SASRec author's transformer blocks architecture but with torch MHA realisation.
```
Comment (blondered) [inline rectools/models/sasrec.py:176]:
```suggestion
    Exactly SASRec author's transformer blocks architecture but with pytorch Multi-Head Attention realisation.
```

File: rectools/models/sasrec.py [hunk 3]
```diff
@@ -111,14 +153,39 @@ def __init__(self, n_factors: int, n_factors_ff: int, dropout_rate: float) -> No
...
+
+    Parameters
+    ----------
+    n_blocks: int
+        Number of self-attention blocks.
```
Comment (blondered) [inline rectools/models/sasrec.py:181]:
```suggestion
        Number of transformer blocks.
```

File: rectools/models/sasrec.py [hunk 4]
```diff
@@ -228,7 +362,32 @@ def forward(self, sessions: torch.Tensor, timeline_mask: torch.Tensor) -> torch.
...
 
 class TransformerBasedSessionEncoder(torch.nn.Module):
-    """TODO"""
+    """
+    Torch model for recommendations.
+
+    Parameters
+    ----------
+    n_blocks: int
+        Number of self-attention blocks.
```
Comment (blondered) [inline rectools/models/sasrec.py:371]:
```suggestion
        Number of transformer blocks.
```

File: rectools/models/sasrec.py [hunk 5]
```diff
@@ -228,7 +362,32 @@ def forward(self, sessions: torch.Tensor, timeline_mask: torch.Tensor) -> torch.
...
+    session_max_len: int
+        Maximum length of user sequence.
+    dropout_rate: float
+        Probability of a hidden unit to be zeroed.
+    use_pos_emb: bool, default True
+        If ``True``, adds learnable positional encoding to session item embeddings.
+    use_causal_attn: bool, default True
+        If ``True``, causal mask is used in multi-head self-attention.
+    transformer_layers_type: Type(TransformerLayersBase), default `SasRecTransformerLayers`
+        Type of transformer layers used for training.
```
Comment (blondered) [inline rectools/models/sasrec.py:385]:
```suggestion
        Type of transformer layers architecture.
```

File: rectools/models/sasrec.py [hunk 6]
```diff
@@ -228,7 +362,32 @@ def forward(self, sessions: torch.Tensor, timeline_mask: torch.Tensor) -> torch.
...
+    item_net_type: Type(ItemNetBase), default IdEmbeddingsItemNet
+        Type of network returning item embeddings.
+    pos_encoding_type: Type(PositionalEncodingBase), default LearnableInversePositionalEncoding
+        Type of the class returning positional embeddings.
```
Comment (blondered) [inline rectools/models/sasrec.py:389]:
```suggestion
        Type of positional encoding.
```

File: rectools/models/sasrec.py [hunk 7]
```diff
@@ -260,17 +419,33 @@ def __init__(
...
+            User sessions consisting of items.
+        item_embs: torch.Tensor
+            Item embeddings.
 
         Returns
         -------
-            torch.Tensor. [batch_size, session_max_len, n_factors]
-
+        torch.Tensor. [batch_size, session_max_len, n_factors]
+            Encoded sessions with positional embeddings that passed transformer blocks.
```
Comment (blondered) [inline rectools/models/sasrec.py:448]:
```suggestion
            Encoded session embeddings.
```

File: rectools/models/sasrec.py [hunk 8]
```diff
@@ -199,14 +308,39 @@ def forward(self, seqs: torch.Tensor, timeline_mask: torch.Tensor, attn_mask: to
...
 
     def forward(self, sessions: torch.Tensor, timeline_mask: torch.Tensor) -> torch.Tensor:
-        """TODO"""
+        """
+        Forward pass to add learnable positional encoding to sessions and mask padding elements.
+
+        Parameters
+        ----------
+        sessions: torch.Tensor
+            User sessions consisting of items.
```
Comment (blondered) [inline rectools/models/sasrec.py:335]:
```suggestion
            User sessions in the form of sequences of items ids.
```

File: rectools/models/sasrec.py [hunk 9]
```diff
@@ -260,17 +419,33 @@ def __init__(
...
         """
-        Pass user history through item embeddings and transformer blocks.
+        Pass user history through item embeddings.
+        Add positional encoding.
+        Pass history through transformer blocks.
+
+        Parameters
+        ----------
+        sessions:  torch.Tensor
+            User sessions consisting of items.
```
Comment (blondered) [inline rectools/models/sasrec.py:441]:
```suggestion
            User sessions in the form of sequences of items ids.
```

File: rectools/models/sasrec.py [hunk 10]
```diff
@@ -199,14 +308,39 @@ def forward(self, seqs: torch.Tensor, timeline_mask: torch.Tensor, attn_mask: to
...
+        ----------
+        sessions: torch.Tensor
+            User sessions consisting of items.
+        timeline_mask: torch.Tensor
+            Mask to zero out padding elements.
+
+        Returns
+        -------
+        torch.Tensor
+            User sessions with positional encoding if use_pos_emb is ``True``.
```
Comment (blondered) [inline rectools/models/sasrec.py:342]:
```suggestion
            Encoded user sessions with added positional encoding if `use_pos_emb` is ``True``.
```

File: rectools/models/sasrec.py [hunk 11]
```diff
@@ -289,7 +464,22 @@ def forward(
...
+        """
+        Forward pass to get logits.
+        Get item embeddings.
+        Pass user sessions through transformer blocks.
+        Calculate logits.
+
+        Parameters
+        ----------
+        sessions: torch.Tensor
+            User sessions consisting of items.
```
Comment (blondered) [inline rectools/models/sasrec.py:476]:
```suggestion
            User sessions in the form of sequences of items ids.
```

File: rectools/models/sasrec.py [hunk 12]
```diff
@@ -300,7 +490,16 @@ def forward(
...
 
 class SequenceDataset(TorchDataset):
-    """TODO"""
+    """
+    Dataset for sequential data.
+
+    Parameters
+    ----------
+    sessions: List[List[int]]
+        User interaction sequences.
```
Comment (blondered) [inline rectools/models/sasrec.py:499]:
```suggestion
        User sessions in the form of sequences of items ids.
```

File: tests/models/test_sasrec.py [hunk 13]
```diff
@@ -0,0 +1,556 @@
...
+from pytorch_lightning import seed_everything
+
+from rectools.columns import Columns
+from rectools.dataset import Dataset, IdMap, Interactions
+from rectools.models.sasrec import SASRecDataPreparator, SASRecModel, SequenceDataset
+from tests.models.utils import assert_second_fit_refits_model
+from tests.testing_utils import assert_id_map_equal, assert_interactions_set_equal
+
+
+@pytest.mark.filterwarnings("ignore::pytorch_lightning.utilities.warnings.PossibleUserWarning")
```
Comment (feldlime) [inline tests/models/test_sasrec.py:17]: Could you please add comments for this and the next line? (What the warnings are and why do we ignore them)

File: tests/models/test_sasrec.py [hunk 14]
```diff
@@ -0,0 +1,556 @@
...
+            session_max_len=3,
+            lr=0.001,
+            batch_size=4,
+            epochs=2,
+            deterministic=True,
+        )
+        model.fit(dataset=dataset)
+        users = np.array([10, 30, 40])
+        actual = model.recommend(users=users, dataset=dataset, k=3, filter_viewed=filter_viewed)
+        actual[Columns.Item] = actual[Columns.Item].apply(int)
```
[the remark this answers is not in the document]
Comment (feldlime) [reply tests/models/test_sasrec.py:94]: So we still expect incorrect id types in reco? Could you explain the reason?

[the remark this answers is not in the document]
Comment (blondered) [reply tests/models/test_sasrec.py:94]: We receive int item ids but pd.Series has dtype object. So the values are correct and they are not float. But dtype is not equal to int.

File: rectools/models/sasrec.py (hunk 1 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 2 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 3 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 4 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 5 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 6 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 7 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 8 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 9 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 10 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 11 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 12 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py:622]: fixed

File: tests/models/test_sasrec.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (spirinamayya) [reply tests/models/test_sasrec.py]: fixed

File: tests/models/test_sasrec.py (hunk 13 above)
Comment (spirinamayya) [reply tests/models/test_sasrec.py:18]: fixed

File: rectools/models/base.py [hunk 15]
```diff
@@ -194,6 +194,10 @@ def recommend(
         reco_warm_final = self._reco_to_external(reco_warm, dataset.user_id_map, dataset.item_id_map)
         reco_cold_final = self._reco_items_to_external(reco_cold, dataset.item_id_map)
 
+        reco_hot_final = self._adjust_reco_types(reco_hot_final, dataset.user_id_map.external_dtype, item_type)
```
Comment (blondered) [inline rectools/models/base.py:197]: Let's just pass it like `dataset.item_id_map.external_dtype` just like we are doing for users

File: rectools/models/sasrec.py (hunk 12 above)
Comment (blondered) [reply rectools/models/sasrec.py:622]: This one missed

File: tests/models/test_sasrec.py [hunk 16]
```diff
@@ -0,0 +1,581 @@
...
+from rectools.dataset import Dataset, IdMap, Interactions
+from rectools.models.sasrec import IdEmbeddingsItemNet, SASRecDataPreparator, SASRecModel, SequenceDataset
+from tests.models.utils import assert_second_fit_refits_model
+from tests.testing_utils import assert_id_map_equal, assert_interactions_set_equal
+
+
+# Ignore pytorch_lightning warnings
+@pytest.mark.filterwarnings("ignore::pytorch_lightning.utilities.warnings.PossibleUserWarning")
+# Ignore non-empty checkpoint directory warnings, SASRec user warnings
+@pytest.mark.filterwarnings("ignore::UserWarning")
```
Comment (blondered) [inline tests/models/test_sasrec.py:20]: You can just change to `on_supported_targets=ignore` where you don't want the warnings to be received from `recommend` method. Let's drop this filtering

**Review by blondered — CHANGES_REQUESTED** [review]

File: tests/models/test_sasrec.py (hunk 13 above)
Comment (feldlime) [reply tests/models/test_sasrec.py:18]: It's quite obvious that it's a pytorch_lightning warning, could you add some details? And also some short explanation why we have this warning and why we ignore it

File: rectools/models/base.py (hunk 15 above)
Comment (spirinamayya) [reply rectools/models/base.py:197]: dataset.item_id_map has type object after _custom_transform_dataset_u2i

File: rectools/models/sasrec.py (hunk 12 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: tests/models/test_sasrec.py (hunk 13 above)
Comment (spirinamayya) [reply tests/models/test_sasrec.py]: removed these filterwarnings. There is one left for class TestSASRecDataPreparator

File: tests/models/test_sasrec.py (hunk 16 above)
Comment (spirinamayya) [reply tests/models/test_sasrec.py]: fixed

**Review by blondered — APPROVED** [review]

**Pull request merged**